File: zig/examples/basic_usage.zig
//! Basic usage example for OpenADR 3 Zig client

const std = @import("std");
const openadr3 = @import("openadr3");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Initialize the client with configuration
    const config = try openadr3.OADR3Config.init(
        "https://your-vtn-server.com",
        "your-client-id",
        "your-client-secret",
    );

    const client_config = config
        .withScope("read_targets write_programs")
        .withTimeout(30);

    var client = try openadr3.OADR3.init(allocator, client_config);
    defer client.deinit();

    std.debug.print("=== OpenADR 3 Zig Client Demo ===\n\n", .{});

    // Search for programs
    std.debug.print("1. Searching for programs...\n", .{});
    const programs = client.searchAllPrograms(null, null, null) catch |err| {
        std.debug.print("Error searching programs: {any}\n", .{err});
        return;
    };
    defer programs.deinit(allocator);

    if (programs.isSuccess()) {
        if (programs.getData()) |program_list| {
            std.debug.print("✅ Found {d} programs\n", .{program_list.len});
            for (program_list, 0..) |program, i| {
                std.debug.print("  {d}. {s} ({s})\n", .{ i + 1, program.program_name, program.id });
                std.debug.print("     Retailer: {s}\n", .{program.retailer_name});
                std.debug.print("     Type: {s}\n", .{program.program_type.toString()});

                // Clean up program data
                program.deinit(allocator);
            }
        }
    } else {
        std.debug.print("❌ Error searching programs\n", .{});
    }

    std.debug.print("\n🎉 Demo completed successfully!\n", .{});
}

File: rust/examples/basic_usage.rs
//! Basic usage example for OpenADR 3 Rust client

use openadr3_client::{OADR3, OADR3Config, DemoProgram, DemoProgramType, DemoTargetType};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize the client with configuration
    let config = OADR3Config::new(
        "https://your-vtn-server.com".to_string(),
        "your-client-id".to_string(),
        "your-client-secret".to_string(),
    )?;

    let client = OADR3::new(config)?;

    // Search for programs
    println!("Searching for programs...");
    let programs = client.search_all_programs(None, None, None).await?;
    
    if programs.is_success() {
        if let Some(program_list) = programs.response() {
            println!("Found {} programs", program_list.len());
            for program in program_list {
                println!("  - {} ({})", program.program_name, program.id);
            }
        }
    } else {
        println!("Error searching programs: {:?}", programs.problem());
    }

    // Create a new program
    println!("\nCreating a new program...");
    let new_program = DemoProgram {
        id: "demo-program-123".to_string(),
        program_name: "Peak Hour Demand Response".to_string(),
        retailer_name: "Example Utility Company".to_string(),
        country: "US".to_string(),
        principal_subdivision: Some("CA".to_string()),
        program_type: DemoProgramType::DemandResponse,
        targets: vec![DemoTargetType::Commercial, DemoTargetType::Industrial],
    };

    let created = client.create_program(new_program).await?;
    if created.is_success() {
        println!("Program created successfully!");
        if let Some(program) = created.response() {
            println!("  ID: {}", program.id);
            println!("  Name: {}", program.program_name);
        }
    } else {
        println!("Error creating program: {:?}", created.problem());
    }

    // Get a specific program
    println!("\nFetching program details...");
    let program_detail = client.get_program("demo-program-123").await?;
    if program_detail.is_success() {
        println!("Program details retrieved successfully!");
    }

    // Search events for the program
    println!("\nSearching for events...");
    let events = client.search_events("demo-program-123", None, Some(10)).await?;
    if events.is_success() {
        if let Some(event_list) = events.response() {
            println!("Found {} events", event_list.len());
        }
    }

    // Search VENs
    println!("\nSearching for VENs...");
    let vens = client.search_vens(None, Some(5)).await?;
    if vens.is_success() {
        if let Some(ven_list) = vens.response() {
            println!("Found {} VENs", ven_list.len());
        }
    }

    println!("\nDemo completed successfully!");
    Ok(())
}